File: src/lib/route-trimming.ts
import { getCity, getQuest } from "./quests";
import { routeTotals } from "./routing";
import type { CityId, TrimmedQuest } from "./types";

export function trimQuestToBudget(cityId: CityId, questId: string, minutes: number): TrimmedQuest {
  const city = getCity(cityId);
  const quest = getQuest(cityId, questId);
  const budget = Math.max(30, Math.min(300, Math.round(minutes)));
  let selected = quest.stops.slice(0, 2);

  for (let count = 2; count <= quest.stops.length; count += 1) {
    const candidate = quest.stops.slice(0, count);
    if (routeTotals(candidate).totalMinutes <= budget) {
      selected = candidate;
    }
  }

  const totals = routeTotals(selected);
  return {
    city,
    quest,
    stops: selected,
    ...totals,
    skippedStops: quest.stops.length - selected.length,
    source: "fallback"
  };
}

File: backend/app/route_trimming.py
from app.data import get_city, get_quest
from app.models import TrimmedQuest
from app.routing import route_totals


def trim_quest_to_budget(city_id: str, quest_id: str, minutes: int) -> TrimmedQuest:
    city = get_city(city_id)
    quest = get_quest(city_id, quest_id)
    budget = max(30, min(300, round(minutes)))
    selected = quest.stops[:2]

    for count in range(2, len(quest.stops) + 1):
        candidate = quest.stops[:count]
        _, _, total_minutes = route_totals(candidate)
        if total_minutes <= budget:
            selected = candidate

    stop_minutes, walk_minutes, total_minutes = route_totals(selected)
    return TrimmedQuest(
        city=city,
        quest=quest,
        stops=selected,
        totalMinutes=total_minutes,
        stopMinutes=stop_minutes,
        walkMinutes=walk_minutes,
        skippedStops=len(quest.stops) - len(selected),
        source="fallback",
    )

